clients/drcachesim/drpt2trace/pt2ir.h

```diff
@@ -342,66 +323,61 @@ class pt2ir_t {
-     * code.
-     * \note The convert function performs two processes:
-     * (1) decode the PT raw trace into libipt's IR format pt_insn;
-     * (2) convert pt_insn into the DynamoRIO's IR format instr_t and append it to ilist.
-     * \note The caller does not need to initialize ilist. But if the convertion is
-     * successful, the caller needs to destroy the ilist.
+     * returns #PT2IR_CONV_SUCCESS. Otherwise, the function returns the corresponding
+     * error code.
+     * \note The convert function performs two processes: (1) decode the PT raw trace into
+     * libipt's IR format pt_insn; (2) convert pt_insn into the DynamoRIO's IR format
+     * instr_t and append it to ilist.
      */
     pt2ir_convert_status_t
-    convert(OUT instrlist_autoclean_t &ilist);
+    convert(IN const uint8_t *next_pt_data, IN size_t next_pt_data_size,
+            INOUT drir_t *drir);
 
 private:
-    /* Load PT raw file to buffer. The struct pt_insn_decoder will decode this buffer to
-     * libipt's IR.
-     */
-    bool
-    load_pt_raw_file(IN std::string &path);
-
-    /* Load the elf section in kcore to sideband session iscache and store the section
-     * index to sideband kimage.
-     * \note XXX: Could we not use kcore? We can store all kernel modules and the mapping
-     * information.
-     */
-    bool
-    load_kcore(IN std::string &path);
-
-    /* Allocate a sideband decoder in the sideband session. The sideband session may
-     * allocate many decoders, which mainly work on handling sideband perf records and
-     * help the PT decoder switch images.
-     */
-    bool
-    alloc_sb_pevent_decoder(IN struct pt_sb_pevent_config &config);
-
     /* Diagnose converting errors and output diagnostic results.
      * It will used to generate the error message during the decoding process.
      */
     void
     dx_decoding_error(IN int errcode, IN const char *errtype, IN uint64_t ip);
 
+    /* It indicate if the pt2ir (pointer-to-intermediate-representation) has been
+     * initialized, signifying the readiness of the conversion process from PT data to
+     * DR's IR.
+     */
+    bool pt2ir_initialized_;
+
     /* Buffer for caching the PT raw trace. */
-    std::unique_ptr<unsigned char> pt_raw_buffer_;
+    std::unique_ptr<uint8_t[]> pt_raw_buffer_;
     size_t pt_raw_buffer_size_;
 
     /* The libipt instruction decoder. */
     struct pt_insn_decoder *pt_instr_decoder_;
 
-    /* The libipt image section cache. */
-    struct pt_image_section_cache *pt_iscache_;
+    /* The libipt sideband image section cache. */
+    struct pt_image_section_cache *pt_sb_iscache_;
 
     /* The libipt sideband session. */
     struct pt_sb_session *pt_sb_session_;
+
+    /* The size of the PT data within the raw buffer. */
+    uint64_t pt_raw_buffer_data_size_;
+
+    /* The offset within the raw buffer of the end of the PT data currently being decoded.
```

Comment:
Add to comment why pt_raw_buffer_data_size_ and cur_pt_data_end_offset_ can be different. This is because of our design to reuse the buffer right? Document more details in comments.